src/persist-client/src/internal/gc.rs

```diff
@@ -190,166 +192,203 @@ where
 
     pub async fn gc_and_truncate(machine: &mut Machine<K, V, T, D>, req: GcReq) {
         assert_eq!(req.shard_id, machine.shard_id());
-        // NB: Because these requests can be processed concurrently (and in
-        // arbitrary order), all of the logic below has to work even if we've
-        // already gc'd and truncated past new_seqno_since.
-
-        let mut states = machine
-            .state_versions
-            .fetch_all_live_states::<K, V, T, D>(&req.shard_id)
-            .await
-            .expect("shard codecs should not change");
-
-        debug!(
-            "gc {} for {} got {} versions from scan",
-            req.shard_id,
-            req.new_seqno_since,
-            states.len()
-        );
-
-        let earliest_live_seqno = match states.peek_seqno() {
-            Some(x) => x,
-            None => panic!(
-                "initialized shard {} should have at least one live state",
-                req.shard_id
-            ),
+
+        let mut next_maintenance = RoutineMaintenance {
+            garbage_collection: Some(req),
+            lease_expiration: None,
+            write_rollup: None,
         };
-        // Fast-path: Someone already GC'd past `req.new_seqno_since`, don't
-        // bother running any of the below logic.
-        //
-        // Also a fix for #14580.
-        if earliest_live_seqno > req.new_seqno_since {
-            machine.metrics.gc.noop.inc();
+
+        while !next_maintenance.is_empty() {
```

Comment:
I think we should be able to move this into `maintenance`, rather than having gc worry about it. We could return `RoutineMaintenance` from this method and pass it back to our caller through the oneshot channel, rather than the `()` marker value we send back now. Then it'll look a lot more like the lease expiry / compaction futures that pass back maintenance upon completion, and we can have a single spot in the code that loops until maintenance is complete.

Reply:
Ah, yes, you're right! That does feel much better... will take a swing at it.